File: src/opencortex/mcp_server.py
# SPDX-License-Identifier: Apache-2.0
"""
MCP Server for OpenCortex.

Exposes MemoryOrchestrator capabilities as MCP tools for external AI agents.
Uses PrefectHQ FastMCP (v3) for tool registration with stdio/SSE/HTTP transport.

Usage::

    # stdio mode (local agent)
    python -m opencortex.mcp_server

    # SSE mode (remote agent)
    python -m opencortex.mcp_server --transport sse --port 8920

    # streamable-http mode
    python -m opencortex.mcp_server --transport streamable-http --port 8920
"""

import argparse
import logging
from contextlib import asynccontextmanager
from typing import Any, Dict, List, Optional

from fastmcp import Context, FastMCP

from opencortex.config import get_config, init_config
from opencortex.orchestrator import MemoryOrchestrator
from opencortex.retrieve.types import ContextType

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Lifespan: initialize orchestrator once at startup
# ---------------------------------------------------------------------------

@asynccontextmanager
async def _lifespan(server: FastMCP):
    """Initialize and teardown the MemoryOrchestrator."""
    config = get_config()
    orch = MemoryOrchestrator(config=config)
    await orch.init()
    logger.info("[MCP] Orchestrator initialized (tenant=%s, user=%s)",
                config.tenant_id, config.user_id)
    try:
        yield {"orchestrator": orch}
    finally:
        await orch.close()
        logger.info("[MCP] Orchestrator closed")


# ---------------------------------------------------------------------------
# Server instance
# ---------------------------------------------------------------------------

mcp = FastMCP(
    name="opencortex",
    instructions=(
        "OpenCortex Memory Server. "
        "Store, search, and manage AI agent memories with SONA reinforcement learning."
    ),
    lifespan=_lifespan,
)


def _get_orch(ctx: Context) -> MemoryOrchestrator:
    """Extract orchestrator from the MCP lifespan context."""
    return ctx.lifespan_context["orchestrator"]


# ---------------------------------------------------------------------------
# Tools
# ---------------------------------------------------------------------------

@mcp.tool(
    name="memory_store",
    description=(
        "Store a new memory, resource, or skill. "
        "Returns the URI and metadata of the stored context."
    ),
)
async def memory_store(
    abstract: str,
    ctx: Context,
    content: str = "",
    category: str = "",
    context_type: str = "memory",
    uri: Optional[str] = None,
    meta: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Store a new context in the memory system.

    Args:
        abstract: Short summary of the memory (used for vector search).
        content: Full content (stored on filesystem as L2).
        category: Category hint (e.g. "preferences", "entities", "patterns").
        context_type: Type of context: "memory", "resource", or "skill".
        uri: Explicit URI. Auto-generated if not provided.
        meta: Additional metadata dict.
    """
    orch = _get_orch(ctx)
    result = await orch.add(
        abstract=abstract,
        content=content,
        category=category,
        context_type=context_type,
        uri=uri,
        meta=meta,
    )
    return {
        "uri": result.uri,
        "context_type": result.context_type,
        "category": result.category,
        "abstract": result.abstract,
    }


@mcp.tool(
    name="memory_search",
    description=(
        "Semantic search across stored memories, resources, and skills. "
        "Returns ranked results with relevance scores."
    ),
)
async def memory_search(
    query: str,
    ctx: Context,
    limit: int = 5,
    context_type: Optional[str] = None,
    category: Optional[str] = None,
) -> Dict[str, Any]:
    """Search for relevant contexts.

    Args:
        query: Natural language search query.
        limit: Maximum number of results (default 5).
        context_type: Restrict to "memory", "resource", or "skill".
        category: Filter by category.
    """
    orch = _get_orch(ctx)

    ct = None
    if context_type:
        ct = ContextType(context_type)

    metadata_filter = None
    if category:
        metadata_filter = {"op": "must", "field": "category", "conds": [category]}

    result = await orch.search(
        query=query,
        limit=limit,
        context_type=ct,
        metadata_filter=metadata_filter,
    )

    items = []
    for matched in result:
        items.append({
            "uri": matched.uri,
            "abstract": matched.abstract,
            "context_type": str(matched.context_type),
            "score": getattr(matched, "score", None),
        })

    return {
        "results": items,
        "total": result.total,
    }


@mcp.tool(
    name="memory_feedback",
    description=(
        "Submit reward feedback for a memory (SONA reinforcement). "
        "Positive rewards reinforce retrieval; negative rewards penalize it."
    ),
)
async def memory_feedback(
    uri: str,
    reward: float,
    ctx: Context,
) -> Dict[str, str]:
    """Submit reward signal for a context.

    Args:
        uri: URI of the context to reward.
        reward: Scalar reward value (positive = good, negative = bad).
    """
    orch = _get_orch(ctx)
    await orch.feedback(uri=uri, reward=reward)
    return {"status": "ok", "uri": uri, "reward": str(reward)}


@mcp.tool(
    name="memory_stats",
    description="Get system statistics including storage info, tenant, and component status.",
)
async def memory_stats(ctx: Context) -> Dict[str, Any]:
    """Return orchestrator statistics."""
    orch = _get_orch(ctx)
    return await orch.stats()


@mcp.tool(
    name="memory_decay",
    description=(
        "Trigger time-decay across all stored memories (SONA). "
        "Reduces effective scores of inactive memories over time."
    ),
)
async def memory_decay(ctx: Context) -> Dict[str, Any]:
    """Trigger time-decay across all records."""
    orch = _get_orch(ctx)
    result = await orch.decay()
    return result or {}


@mcp.tool(
    name="memory_health",
    description="Check health status of all OpenCortex components.",
)
async def memory_health(ctx: Context) -> Dict[str, Any]:
    """Check health of all components."""
    orch = _get_orch(ctx)
    return await orch.health_check()


# ---------------------------------------------------------------------------
# RuVector Hooks Tools (Native Self-Learning)
# ---------------------------------------------------------------------------

@mcp.tool(
    name="memory_hooks_learn",
    description=(
        "Record a learning outcome using RuVector native Q-learning. "
        "Maps OpenCortex concepts to hooks: state=URI, action=context_type, reward=feedback. "
        "Returns best action recommendation based on learned patterns."
    ),
)
async def memory_hooks_learn(
    state: str,
    action: str,
    reward: float,
    available_actions: str = "",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Record a learning outcome via RuVector hooks Q-learning.

    Args:
        state: Current state (e.g., URI, query context).
        action: Action taken (e.g., "memory", "skill", "resource").
        reward: Reward value (-1 to 1).
        available_actions: Comma-separated available actions.
    """
    orch = _get_orch(ctx)
    actions = available_actions.split(",") if available_actions else None
    return await orch.hooks_learn(
        state=state,
        action=action,
        reward=reward,
        available_actions=actions,
    )


@mcp.tool(
    name="memory_hooks_remember",
    description=(
        "Store content in RuVector semantic memory. "
        "Useful for remembering important context that should persist beyond session."
    ),
)
async def memory_hooks_remember(
    content: str,
    memory_type: str = "general",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Store content in semantic memory.

    Args:
        content: Content to remember.
        memory_type: Type of memory (default: general).
    """
    orch = _get_orch(ctx)
    return await orch.hooks_remember(content=content, memory_type=memory_type)


@mcp.tool(
    name="memory_hooks_recall",
    description=(
        "Search RuVector semantic memory for relevant content. "
        "Different from vector search - searches learned patterns and memories."
    ),
)
async def memory_hooks_recall(
    query: str,
    limit: int = 5,
    ctx: Context = None,
) -> List[Dict[str, Any]]:
    """Search semantic memory.

    Args:
        query: Search query.
        limit: Maximum results.
    """
    orch = _get_orch(ctx)
    return await orch.hooks_recall(query=query, limit=limit)


@mcp.tool(
    name="memory_hooks_trajectory_begin",
    description="Begin tracking a learning trajectory for multi-step tasks.",
)
async def memory_hooks_trajectory_begin(
    trajectory_id: str,
    initial_state: str,
    ctx: Context = None,
) -> Dict[str, Any]:
    """Begin a learning trajectory."""
    orch = _get_orch(ctx)
    return await orch.hooks_trajectory_begin(
        trajectory_id=trajectory_id,
        initial_state=initial_state,
    )


@mcp.tool(
    name="memory_hooks_trajectory_step",
    description="Add a step to an existing learning trajectory.",
)
async def memory_hooks_trajectory_step(
    trajectory_id: str,
    action: str,
    reward: float,
    next_state: str = "",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Add a step to trajectory."""
    orch = _get_orch(ctx)
    return await orch.hooks_trajectory_step(
        trajectory_id=trajectory_id,
        action=action,
        reward=reward,
        next_state=next_state or None,
    )


@mcp.tool(
    name="memory_hooks_trajectory_end",
    description="End a learning trajectory with a quality score.",
)
async def memory_hooks_trajectory_end(
    trajectory_id: str,
    quality_score: float,
    ctx: Context = None,
) -> Dict[str, Any]:
    """End a trajectory."""
    orch = _get_orch(ctx)
    return await orch.hooks_trajectory_end(
        trajectory_id=trajectory_id,
        quality_score=quality_score,
    )


@mcp.tool(
    name="memory_hooks_error_record",
    description=(
        "Record an error and its fix for the system to learn from. "
        "Helps the system remember how to fix common errors."
    ),
)
async def memory_hooks_error_record(
    error: str,
    fix: str,
    context: str = "",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Record an error and fix for learning."""
    orch = _get_orch(ctx)
    return await orch.hooks_error_record(
        error=error,
        fix=fix,
        context=context or None,
    )


@mcp.tool(
    name="memory_hooks_error_suggest",
    description=(
        "Get suggested fixes for an error based on learned patterns. "
        "The system will recommend fixes based on previously recorded errors."
    ),
)
async def memory_hooks_error_suggest(
    error: str,
    ctx: Context = None,
) -> List[Dict[str, Any]]:
    """Get suggested fixes for an error."""
    orch = _get_orch(ctx)
    return await orch.hooks_error_suggest(error=error)


@mcp.tool(
    name="memory_hooks_stats",
    description="Get RuVector hooks intelligence statistics (Q-learning patterns, memories, trajectories, errors).",
)
async def memory_hooks_stats(ctx: Context) -> Dict[str, Any]:
    """Get hooks statistics."""
    orch = _get_orch(ctx)
    return await orch.hooks_stats()


# ---------------------------------------------------------------------------
# Session Management Tools (Context Self-Iteration)
# ---------------------------------------------------------------------------

@mcp.tool(
    name="session_begin",
    description=(
        "Begin a new session for context self-iteration. "
        "The session will buffer messages and extract persistent memories on end."
    ),
)
async def session_begin(
    session_id: str,
    ctx: Context,
) -> Dict[str, Any]:
    """Begin a new session.

    Args:
        session_id: Unique session identifier.
    """
    orch = _get_orch(ctx)
    return await orch.session_begin(session_id=session_id)


@mcp.tool(
    name="session_message",
    description=(
        "Add a message to an active session. "
        "Messages are buffered for memory extraction when the session ends."
    ),
)
async def session_message(
    session_id: str,
    role: str,
    content: str,
    ctx: Context,
) -> Dict[str, Any]:
    """Add a message to a session.

    Args:
        session_id: Session identifier.
        role: Message role ("user" or "assistant").
        content: Message content.
    """
    orch = _get_orch(ctx)
    return await orch.session_message(
        session_id=session_id,
        role=role,
        content=content,
    )


@mcp.tool(
    name="session_end",
    description=(
        "End a session and trigger memory extraction. "
        "The system will analyze the conversation and automatically extract "
        "persistent memories (preferences, patterns, skills, errors)."
    ),
)
async def session_end(
    session_id: str,
    quality_score: float = 0.5,
    ctx: Context = None,
) -> Dict[str, Any]:
    """End a session and extract memories.

    Args:
        session_id: Session to end.
        quality_score: Overall session quality (0.0-1.0).
    """
    orch = _get_orch(ctx)
    return await orch.session_end(
        session_id=session_id,
        quality_score=quality_score,
    )


# ---------------------------------------------------------------------------
# Hooks Integration Tools (Route through MCP → Viking FS → RuVector)
# ---------------------------------------------------------------------------

@mcp.tool(
    name="hooks_route",
    description=(
        "Route a task to the best agent based on learned patterns. "
        "Returns the recommended agent and reasoning."
    ),
)
async def hooks_route(
    task: str,
    agents: str = "",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Route task to best agent.

    Args:
        task: Task description or command.
        agents: Comma-separated available agent names.
    """
    orch = _get_orch(ctx)
    agent_list = [a.strip() for a in agents.split(",") if a.strip()] if agents else None
    return await orch.hooks_route(task=task, agents=agent_list)


@mcp.tool(
    name="hooks_init",
    description="Initialize OpenCortex hooks configuration for a project.",
)
async def hooks_init(
    project_path: str = ".",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Initialize hooks for a project.

    Args:
        project_path: Project root path.
    """
    orch = _get_orch(ctx)
    return await orch.hooks_init(project_path=project_path)


@mcp.tool(
    name="hooks_pretrain",
    description="Pre-train OpenCortex from repository content (files, patterns, structure).",
)
async def hooks_pretrain(
    repo_path: str = ".",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Pre-train from repository.

    Args:
        repo_path: Repository root path.
    """
    orch = _get_orch(ctx)
    return await orch.hooks_pretrain(repo_path=repo_path)


@mcp.tool(
    name="hooks_verify",
    description="Verify OpenCortex hooks configuration is correct and functional.",
)
async def hooks_verify(ctx: Context = None) -> Dict[str, Any]:
    """Verify hooks configuration."""
    orch = _get_orch(ctx)
    return await orch.hooks_verify()


@mcp.tool(
    name="hooks_doctor",
    description="Diagnose OpenCortex system health, configuration issues, and connectivity.",
)
async def hooks_doctor(ctx: Context = None) -> Dict[str, Any]:
    """Run diagnostics."""
    orch = _get_orch(ctx)
    return await orch.hooks_doctor()


@mcp.tool(
    name="hooks_export",
    description="Export OpenCortex intelligence data (learned patterns, memories, trajectories).",
)
async def hooks_export(
    format: str = "json",
    ctx: Context = None,
) -> Dict[str, Any]:
    """Export intelligence data.

    Args:
        format: Export format ("json" or "markdown").
    """
    orch = _get_orch(ctx)
    return await orch.hooks_export(format=format)


@mcp.tool(
    name="hooks_build_agents",
    description="Generate agent configuration based on learned patterns and project structure.",
)
async def hooks_build_agents(ctx: Context = None) -> Dict[str, Any]:
    """Generate agent configurations."""
    orch = _get_orch(ctx)
    return await orch.hooks_build_agents()


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(
        prog="opencortex.mcp_server",
        description="OpenCortex MCP Server",
    )
    parser.add_argument(
        "--transport",
        choices=["stdio", "sse", "streamable-http"],
        default="streamable-http",
        help="Transport mode (default: streamable-http)",
    )
    parser.add_argument(
        "--port",
        type=int,
        default=8920,
        help="Port for SSE/HTTP transport (default: 8920)",
    )
    parser.add_argument(
        "--host",
        default="127.0.0.1",
        help="Host for SSE/HTTP transport (default: 127.0.0.1)",
    )
    parser.add_argument(
        "--config",
        default=None,
        help="Path to opencortex.json config file",
    )
    parser.add_argument(
        "--log-level",
        default="INFO",
        choices=["DEBUG", "INFO", "WARNING", "ERROR"],
        help="Log level (default: INFO)",
    )
    parser.add_argument(
        "--stateless",
        action="store_true",
        default=True,
        help="Enable stateless HTTP mode (default: True)",
    )
    parser.add_argument(
        "--json-response",
        action="store_true",
        default=True,
        help="Return JSON responses instead of SSE streams (default: True)",
    )
    args = parser.parse_args()

    logging.basicConfig(
        level=getattr(logging, args.log_level),
        format="%(asctime)s [%(name)s] %(levelname)s: %(message)s",
    )

    # Initialize config
    init_config(path=args.config)

    run_kwargs = {"transport": args.transport}
    if args.transport != "stdio":
        run_kwargs["host"] = args.host
        run_kwargs["port"] = args.port
    if args.stateless and args.transport == "streamable-http":
        run_kwargs["stateless_http"] = True
    if args.json_response and args.transport == "streamable-http":
        run_kwargs["json_response"] = True
    mcp.run(**run_kwargs)


if __name__ == "__main__":
    main()

File: plugins/opencortex-memory/lib/mcp-server.mjs
#!/usr/bin/env node
/**
 * OpenCortex MCP Server — pure Node.js, stdio transport.
 * Thin proxy: MCP JSON-RPC ↔ HTTP REST API.
 * Zero external dependencies.
 */
import { getHttpUrl, ensureDefaultConfig } from './common.mjs';
import { buildClientHeaders } from './http-client.mjs';

// ── Tool definitions ───────────────────────────────────────────────────
// Each entry: [httpMethod, httpPath, description, parameters]
// parameters: { name: { type, description, required?, default? } }

const TOOLS = {
  // ── Core Memory ──
  memory_store: ['POST', '/api/v1/memory/store',
    'Persist a piece of knowledge the user wants remembered across sessions. '
    + 'Use when the user explicitly shares a preference, fact, decision, or correction — '
    + 'NOT for recording conversation turns (use memory_context commit for that). '
    + 'Semantic dedup is on by default: if a similar memory exists, it will be merged instead of duplicated. '
    + 'Returns {uri, context_type, category, abstract, dedup_action?}.', {
      abstract:     { type: 'string',  description: 'One-sentence summary capturing the key point (used for retrieval ranking)', required: true },
      content:      { type: 'string',  description: 'Full detailed content. If >500 chars, the system auto-generates a structured overview from it', default: '' },
      category:     { type: 'string',  description: 'Semantic category. Choose the most specific: profile | preferences | entities | events | cases | patterns | error_fixes | workflows | strategies | documents | plans', default: '' },
      context_type: { type: 'string',  description: 'Storage type: memory (default, for knowledge/facts) | resource (reference docs) | skill (reusable procedures)', default: 'memory' },
      meta:         { type: 'object',  description: 'Arbitrary key-value metadata (e.g. {source: "user", language: "zh"})' },
      dedup:        { type: 'boolean', description: 'Enable semantic dedup — merges into existing similar memory if found. Set false only for intentional duplicates', default: true },
    }],
  memory_batch_store: ['POST', '/api/v1/memory/batch_store',
    'Import multiple documents in one call. Use for bulk ingestion of files, notes, or scan results. '
    + 'Each item is stored independently with its own URI. '
    + 'Returns {stored, skipped, errors}.', {
      items:       { type: 'array',  description: 'Array of objects, each with: {abstract (required), content, category, context_type, meta}', required: true },
      source_path: { type: 'string', description: 'Source directory path for provenance tracking', default: '' },
      scan_meta:   { type: 'object', description: 'Import metadata: {total_files, has_git, project_id}' },
    }],
  memory_search: ['POST', '/api/v1/memory/search',
    'Search stored memories by natural language query. Uses intent-aware retrieval: '
    + 'the system analyzes your query to determine search strategy (top_k, detail level, reranking). '
    + 'Returns {results: [{uri, abstract, overview?, content?, context_type, score}], total}. '
    + 'Use when you need to recall facts, preferences, past decisions, or any previously stored knowledge.', {
      query:        { type: 'string',  description: 'Natural language query describing what you need to recall', required: true },
      limit:        { type: 'integer', description: 'Max results (system may return fewer based on relevance)', default: 5 },
      context_type: { type: 'string',  description: 'Restrict to type: memory | resource | skill. Omit to search all types' },
      category:     { type: 'string',  description: 'Restrict to category (e.g. "preferences", "error_fixes"). Omit to search all categories' },
    }],
  memory_feedback: ['POST', '/api/v1/memory/feedback',
    'Reinforce or penalize a memory via reward signal. Call with positive reward (+0.1 to +1.0) '
    + 'when a retrieved memory was useful. Call with negative reward (-0.1 to -1.0) when it was '
    + 'irrelevant or wrong. This adjusts future retrieval ranking through reinforcement learning.', {
      uri:    { type: 'string', description: 'The opencortex:// URI of the memory to reward (from search results)', required: true },
      reward: { type: 'number', description: 'Reward signal: positive reinforces retrieval, negative penalizes. Typical range: -1.0 to +1.0', required: true },
    }],
  memory_decay: ['POST', '/api/v1/memory/decay',
    'Maintenance: apply time-decay to all memories, reducing scores of inactive ones. '
    + 'Call periodically (e.g. daily) to let unused memories naturally fade. '
    + 'Frequently accessed memories resist decay.', {}],
  system_status: ['GET', '/api/v1/system/status',
    'Check system health and diagnostics. Returns memory count, storage stats, and component status.', {
      type: { type: 'string', description: 'Report depth: health (quick liveness) | stats (counts and sizes) | doctor (full diagnostic)', default: 'doctor' },
    }],

  // ── Session (low-level) ──
  // Prefer memory_context for most use cases — it handles session lifecycle automatically.
  session_begin: ['POST', '/api/v1/session/begin',
    'Low-level: start session recording. Prefer memory_context which auto-creates sessions. '
    + 'Only use directly if you need explicit session control without the context lifecycle.', {
      session_id: { type: 'string', description: 'Unique session identifier', required: true },
    }],
  session_message: ['POST', '/api/v1/session/message',
    'Low-level: record a single message to an active session. '
    + 'Prefer memory_context commit which records the full turn and handles idempotency.', {
      session_id: { type: 'string', description: 'Session identifier (must call session_begin first)', required: true },
      role:       { type: 'string', description: 'Message role: user | assistant', required: true },
      content:    { type: 'string', description: 'Message text content', required: true },
    }],
  session_end: ['POST', '/api/v1/session/end',
    'Low-level: end session and trigger knowledge extraction pipeline. '
    + 'Prefer memory_context with phase="end". '
    + 'The system splits the conversation into task traces and extracts reusable knowledge.', {
      session_id:    { type: 'string', description: 'Session identifier to close', required: true },
      quality_score: { type: 'number', description: 'Overall session quality (0.0-1.0). Higher scores prioritize knowledge extraction', default: 0.5 },
    }],

  // ── Context Protocol (recommended) ──
  memory_context: ['POST', '/api/v1/context',
    'Primary tool for memory-augmented conversations. Manages the full lifecycle in three phases:\n'
    + '\n'
    + 'PHASE 1 — prepare: Call BEFORE generating your response. Retrieves relevant memories and knowledge '
    + 'based on the user\'s message. Returns {memory: [...], knowledge: [...], instructions, intent}. '
    + 'Use the returned context to inform your response. Session is auto-created if needed.\n'
    + '\n'
    + 'PHASE 2 — commit: Call AFTER generating your response. Records the full conversation turn '
    + '(user message + your response). Pass cited_uris to reward memories you actually used. '
    + 'Returns {accepted, write_status, session_turns}. Idempotent — safe to retry.\n'
    + '\n'
    + 'PHASE 3 — end: Call when the conversation is over. Triggers knowledge extraction from the '
    + 'session transcript. Returns {status: "closed", total_turns}.\n'
    + '\n'
    + 'Typical flow per turn: prepare → [generate response] → commit. Call end once at session close.', {
      session_id: { type: 'string', description: 'Stable session identifier — reuse across all turns in one conversation. Alphanumeric, hyphens, underscores, 1-128 chars', required: true },
      phase:      { type: 'string', description: 'Lifecycle phase: "prepare" (before response) | "commit" (after response) | "end" (session close)', required: true },
      turn_id:    { type: 'string', description: 'Unique per-turn ID for idempotency. Required for prepare and commit. Use a counter (t1, t2...) or UUID' },
      messages:   { type: 'array',  description: 'Array of {role, content}. prepare: pass [user message]. commit: pass [user message, assistant response]. Not needed for end' },
      cited_uris: { type: 'array',  description: 'commit only: array of opencortex:// URIs from prepare results that you referenced in your response. Triggers +0.1 RL reward per URI' },
      config:     { type: 'object', description: 'prepare only: {max_items: 1-20 (default 5), detail_level: "l0"|"l1"|"l2" (default "l1"), recall_mode: "auto"|"always"|"never" (default "auto")}' },
    }],
};

// ── Build JSON Schema for tools/list ───────────────────────────────────
function buildToolSchema(name, [, , description, params]) {
  const properties = {};
  const required = [];
  for (const [pName, pDef] of Object.entries(params)) {
    const prop = { type: pDef.type, description: pDef.description };
    if (pDef.default !== undefined) prop.default = pDef.default;
    properties[pName] = prop;
    if (pDef.required) required.push(pName);
  }
  const schema = { type: 'object', properties };
  if (required.length) schema.required = required;
  return { name, description, inputSchema: schema };
}

// ── HTTP proxy ─────────────────────────────────────────────────────────
const HTTP_URL = getHttpUrl();

async function callTool(name, args) {
  const def = TOOLS[name];
  if (!def) throw new Error(`Unknown tool: ${name}`);
  const [method, path] = def;
  let url = `${HTTP_URL}${path}`;

  // Apply defaults
  const params = def[3];
  const body = {};
  for (const [pName, pDef] of Object.entries(params)) {
    if (args[pName] !== undefined) {
      body[pName] = args[pName];
    } else if (pDef.default !== undefined) {
      body[pName] = pDef.default;
    }
  }

  // Build headers with identity from MCP config
  const hdrs = buildClientHeaders();

  const opts = { method, signal: AbortSignal.timeout(30000) };
  if (method === 'POST') {
    hdrs['Content-Type'] = 'application/json';
    opts.body = JSON.stringify(body);
  } else if (method === 'GET' && Object.keys(body).length > 0) {
    const qs = new URLSearchParams(body).toString();
    url = `${url}?${qs}`;
  }
  opts.headers = hdrs;

  const res = await fetch(url, opts);
  const text = await res.text();
  if (!res.ok) throw new Error(`${method} ${path} → ${res.status}: ${text}`);
  try { return JSON.parse(text); } catch { return text; }
}

// ── JSON-RPC stdio transport ───────────────────────────────────────────
function send(msg) {
  const json = JSON.stringify(msg);
  process.stdout.write(`${json}\n`);
}

function jsonrpcResult(id, result) {
  send({ jsonrpc: '2.0', id, result });
}

function jsonrpcError(id, code, message) {
  send({ jsonrpc: '2.0', id, error: { code, message } });
}

async function handleMessage(msg) {
  const { id, method, params } = msg;

  switch (method) {
    case 'initialize':
      return jsonrpcResult(id, {
        protocolVersion: '2024-11-05',
        capabilities: { tools: {} },
        serverInfo: { name: 'opencortex', version: '0.4.2' },
      });

    case 'notifications/initialized':
      return; // no response needed

    case 'tools/list':
      return jsonrpcResult(id, {
        tools: Object.entries(TOOLS).map(([name, def]) => buildToolSchema(name, def)),
      });

    case 'tools/call': {
      const toolName = params?.name;
      const toolArgs = params?.arguments || {};
      try {
        const result = await callTool(toolName, toolArgs);
        return jsonrpcResult(id, {
          content: [{ type: 'text', text: typeof result === 'string' ? result : JSON.stringify(result) }],
        });
      } catch (err) {
        return jsonrpcResult(id, {
          content: [{ type: 'text', text: `Error: ${err.message}` }],
          isError: true,
        });
      }
    }

    default:
      if (id != null) {
        jsonrpcError(id, -32601, `Method not found: ${method}`);
      }
  }
}

// ── Main loop ──────────────────────────────────────────────────────────
async function main() {
  ensureDefaultConfig();
  let buffer = '';
  for await (const chunk of process.stdin) {
    buffer += chunk.toString();
    // Process complete lines
    let nl;
    while ((nl = buffer.indexOf('\n')) !== -1) {
      const line = buffer.slice(0, nl).trim();
      buffer = buffer.slice(nl + 1);
      if (!line) continue;
      try {
        const msg = JSON.parse(line);
        await handleMessage(msg);
      } catch (err) {
        process.stderr.write(`[opencortex-mcp] parse error: ${err.message}\n`);
      }
    }
  }
}

main();
